src/common/hist_util.cc

```diff
@@ -419,28 +372,36 @@ void GHistBuilder<GradientSumT>::BuildHist(
 
   if (contiguousBlock) {
     // contiguous memory access, built-in HW prefetching is enough
-    BuildHistKernel<GradientSumT, false>(gpair, row_indices, gmat, isDense, hist);
+    BuildHistDispatch<GradientSumT, false, any_missing>(gpair, row_indices, gmat, hist);
   } else {
     const RowSetCollection::Elem span1(row_indices.begin, row_indices.end - no_prefetch_size);
     const RowSetCollection::Elem span2(row_indices.end - no_prefetch_size, row_indices.end);
 
-    BuildHistKernel<GradientSumT, true>(gpair, span1, gmat, isDense, hist);
+    BuildHistDispatch<GradientSumT, true, any_missing>(gpair, span1, gmat, hist);
     // no prefetching to avoid loading extra memory
-    BuildHistKernel<GradientSumT, false>(gpair, span2, gmat, isDense, hist);
+    BuildHistDispatch<GradientSumT, false, any_missing>(gpair, span2, gmat, hist);
   }
 }
 template
-void GHistBuilder<float>::BuildHist(const std::vector<GradientPair>& gpair,
+void GHistBuilder<float>::BuildHist<true>(const std::vector<GradientPair>& gpair,
+                             const RowSetCollection::Elem row_indices,
+                             const GHistIndexMatrix& gmat,
+                             GHistRow<float> hist);
+template
+void GHistBuilder<float>::BuildHist<false>(const std::vector<GradientPair>& gpair,
+                             const RowSetCollection::Elem row_indices,
+                             const GHistIndexMatrix& gmat,
+                             GHistRow<float> hist);
+template
+void GHistBuilder<double>::BuildHist<true>(const std::vector<GradientPair>& gpair,
                              const RowSetCollection::Elem row_indices,
                              const GHistIndexMatrix& gmat,
-                             GHistRow<float> hist,
-                             bool isDense);
+                             GHistRow<double> hist);
 template
-void GHistBuilder<double>::BuildHist(const std::vector<GradientPair>& gpair,
+void GHistBuilder<double>::BuildHist<false>(const std::vector<GradientPair>& gpair,
```

Comment:
I don't think you need to template these functions.  The `GHistIndexMatrix` knows whether data contains missing values: https://github.com/dmlc/xgboost/blob/72f9daf9b6f29a4ff9ac23ac23ecd1c57936be93/src/common/hist_util.h#L288

Reply:
It was done for reducing branching in `hist_util.cc`  (as there is already branching related to `BinIdxType` ). And as `any_missing` was set above we avoid explicit branching for partition also.